File: src/test/java/com/github/nsc/de/compiler/parser/ParserTestUtil.java
package com.github.nsc.de.compiler.parser;

import com.github.nsc.de.compiler.lexer.Lexer;
import com.github.nsc.de.compiler.lexer.characterinputstream.CharacterInputStream;
import com.github.nsc.de.compiler.lexer.characterinputstream.StringCharacterInputStream;
import com.github.nsc.de.compiler.lexer.token.TokenInputStream;
import com.github.nsc.de.compiler.parser.node.Node;
import com.github.nsc.de.compiler.parser.node.Tree;
import com.github.nsc.de.compiler.parser.parser.Parser;

import static org.junit.jupiter.api.Assertions.*;
import static com.github.nsc.de.compiler.TestUtil.*;

public class ParserTestUtil {

    public static Tree parse(String source, String input) {

        CharacterInputStream in = new StringCharacterInputStream(source, input);
        Lexer lexer = new Lexer(in);
        TokenInputStream tokens = lexer.makeTokens();
        Parser parser = new Parser(tokens);

        return parser.parse();

    }

    @SuppressWarnings("unchecked")
    public static  <T extends Node> T parseSingle(String source, String input, Class<T> type) {

        Tree tree = parse(source, input);
        Node[] nodes = tree.getChildren();

        // Result should be 1 node
        assertEquals(1, nodes.length);

        // result is expected class
        assertNotNull(nodes[0]);
        assertType(type, nodes[0]);

        return (T) nodes[0];

    }
}

File: parser/src/commonTest/kotlin/com/github/shakelang/shake/parser/ParserTestUtil.kt
package com.github.shakelang.shake.parser

import com.github.shakelang.shake.assertType
import com.github.shakelang.shake.lexer.Lexer
import com.github.shakelang.shake.parser.node.Node
import com.github.shakelang.shake.parser.node.Tree
import com.github.shakelang.shake.util.characterinput.characterinputstream.CharacterInputStream
import com.github.shakelang.shake.util.characterinput.characterinputstream.SourceCharacterInputStream
import kotlin.reflect.KClass
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

object ParserTestUtil {
    fun parse(source: String, input: String): Tree {
        val input: CharacterInputStream = SourceCharacterInputStream(source, input)
        val lexer = Lexer(input)
        val tokens = lexer.makeTokens()
        val parser = Parser(tokens)
        return parser.parse()
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : Node> parseSingle(source: String, input: String, type: KClass<T>): T {
        val tree = parse(source, input)
        val nodes = tree.children

        // Result should be 1 node
        assertEquals(1, nodes.size)

        // result is expected class
        assertNotNull(nodes[0])
        assertType(type, nodes[0])
        return nodes[0] as T
    }
}